src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+///
+/// "Crs" stands for "compressed row sparse."  This is the phrase
+/// Trilinos traditionally uses to describe compressed sparse row
+/// storage for sparse matrices, as described, for example, in Saad
+/// (2nd ed.).
+template<class ScalarType,
+         class OrdinalType,
+         class Device,
+         class MemoryTraits = void,
+         class SizeType = typename Kokkos::ViewTraits<OrdinalType*, Device, void, void>::size_type>
+class BlockCrsMatrix {
+private:
+  typedef typename Kokkos::ViewTraits<ScalarType*,Device,void,void>::host_mirror_space host_mirror_space ;
+public:
+  //! Type of the matrix's execution space.
+  typedef typename Device::execution_space execution_space;
+  //! Type of the matrix's memory space.
+  typedef typename Device::memory_space memory_space;
+  //! Type of the matrix's device type.
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+
+  //! Type of each value in the matrix.
+  typedef ScalarType value_type;
+  //! Type of each (column) index in the matrix.
+  typedef OrdinalType ordinal_type;
+  typedef MemoryTraits memory_traits;
+  /// \brief Type of each entry of the "row map."
+  ///
+  /// The "row map" corresponds to the \c ptr array of row offsets in
+  /// compressed sparse row (CSR) storage.
+  typedef SizeType size_type;
+
+  //! Type of a host-memory mirror of the sparse matrix.
+  typedef BlockCrsMatrix<ScalarType, OrdinalType, host_mirror_space, MemoryTraits> HostMirror;
+  //! Type of the graph structure of the sparse matrix.
+  typedef Kokkos::StaticCrsGraph<OrdinalType, Kokkos::LayoutLeft, execution_space, SizeType> StaticCrsGraphType;
+  //! Type of column indices in the sparse matrix.
+  typedef typename StaticCrsGraphType::entries_type index_type;
+  //! Const version of the type of column indices in the sparse matrix.
+  typedef typename index_type::const_value_type const_ordinal_type;
+  //! Nonconst version of the type of column indices in the sparse matrix.
+  typedef typename index_type::non_const_value_type non_const_ordinal_type;
+  //! Type of the "row map" (which contains the offset for each row's data).
+  typedef typename StaticCrsGraphType::row_map_type row_map_type;
+  //! Const version of the type of row offsets in the sparse matrix.
+  typedef typename row_map_type::const_value_type const_size_type;
+  //! Nonconst version of the type of row offsets in the sparse matrix.
+  typedef typename row_map_type::non_const_value_type non_const_size_type;
+  //! Kokkos Array type of the entries (values) in the sparse matrix.
+  typedef Kokkos::View<value_type*, Kokkos::LayoutRight, device_type, MemoryTraits> values_type;
+  //! Const version of the type of the entries in the sparse matrix.
+  typedef typename values_type::const_value_type const_value_type;
+  //! Nonconst version of the type of the entries in the sparse matrix.
+  typedef typename values_type::non_const_value_type non_const_value_type;
+
+#ifdef KOKKOS_USE_CUSPARSE
```

Comment:
Is this just a copy-and-paste from `KokkosSparse::CrsMatrix` or do you really intend to use the cuSPARSE stuff here?

Reply:
Yes, I used `KokkosSparse::CrsMatrix` as the starting point for the code and tried to keep things looking similar for familiarity. I removed all the cuSparse stuff in recent commit.